rfcs/2022-12-09-541-reload-api.md

```diff
@@ -0,0 +1,208 @@
+- Allowing configuration error messages to be passed back via the API
+
+### Out of scope
+
+- Changing the actual reloading semantics of Vector in any way
+- Implementing APIs for any other type of control actions
+
+## Pain
+
+Our enterprise control plane component needs to initate configuration reloads
+in Vector and report back success or failure. Currently, the only way to do
+that would be to manipulate the config files on disk, ensure Vector is watching
+for changes, and parse the log output to determine whether or not the reload
+succeeded. This would be difficult to implement and likely unreliable.
+
+## Proposal
+
+### User Experience
+
+At startup, Vector will look for an environment variable
+(`VECTOR_CONTROL_SOCKET_PATH`). If that variable is present and contains a
+valid path, it will start an HTTP control server listening on a Unix socket at
+that path.
+
+The control server will support a single route at `POST /config`, which will
+accept a JSON-encoded version of a Vector configuration. Similar to config
+received from a provider implementation, input globs will be expanded but
+environment variables will not be interpolated. The HTTP response will
+synchronously indicate the success or failure of the triggered reload,
+including any error messages scoped by component name.
+
+### Implementation
+
+The goal will be to implement a very simple, local-only API via which external
+processes can hook directly into Vector's reloading process. An important
+difference with existing solutions like providers is that the interaction will
+be synchronous, i.e. there will be a response with either success or failure
+and error messages for each reload attempt.
+
+Currently, all of Vectors reloading ultimately happens in the main run loop,
+where we watch for events like a signal, crash of a component, etc, and execute
+the corresponding changes to the topology (slightly different depending on the
+source of the signal). The first challenge will be to consolidate the logic of
+what needs to happen when a config changes (e.g. updating GraphQL API state,
+enterprise reporting, internal events, etc). A rough initial pass at that
+process is included with the RFC PR to help show dependencies. The next step
+would likely be to extract a `TopologyController` or similar struct that owns
+the topology and everything that needs to change along with it. That struct can
+then present a simple API for reloads that ensures everything is done in a
+consistent matter, no matter the source.
+
+Once that struct is in place, it can be wrapped in an `Arc<Mutex<_>>` or
+similar such that it can be shared across both the existing `Signal`-focused
+run loop and a new control server. The new control server itself can be a very
+simple task spawned prior to entering the main loop. As a local-only control
+API, it can different from most HTTP servers in a few ways:
+
+- By listening on a Unix socket instead of the network, it can rely on
```

Comment:
Will this socket approach work on Windows too? It's less of an immediate concern, just wondering if we'll make it harder for ourselves later. Maybe it could work over both a Unix socket _and_ TCP/HTTP?

Reply:
Since Vector does support windows today, we should have parity here indeed - unless we want to intentionally limit support to platforms/deprecate/secondary target/etc.

Reply:
Fun fact, Windows actually [supports Unix sockets](https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/).